### ISSUE TITLE
Painless: Failed resolution of subclasses of functional interfaces

### ISSUE THREAD

**Issue by stu-elastic** [elastic/elasticsearch#81696]
After 7.15, class structure lookup was changed.  This impacted functional
interfaces that define apply in their super classes such as `BinaryOperator`.

Reproduction:
```
[1, 2].stream().reduce((i1, i2) -> i1).orElse(0)
```

Will trigger an `illegal_argument_exception` due to `Cannot pass lambda to [java.util.function.BinaryOperator], not a functional interface`.

From: https://discuss.elastic.co/t/regression-painless-script-fails-for-binaryoperator-lamdba/291714

**Comment by megglos** [comment]
Thanks for verifying so fast, not sure if it is helpful but in the meantime I created a PR with a potential fix see #81697

**Comment by TheFireCookie** [comment]
Hello, I think I encounter the same issue.

I have an index with the following mapping:
```json
            "properties": {
                "folders": {
                    "type": "text",
                    "fields": {
                        "keyword": {
                            "type": "keyword",
                            "ignore_above": 256
                        }
                    }
                }
}
```

I have the following script:
```json
{
    "script": {
        "lang": "painless",
        "source": " ctx._source.folders.replaceAll(f -> f.replace(params.oldFolderPath, params.newFolderPath)); "
    }
}
```

This script was working in 7.15.x and is not working in 7.16.x with the following error:
```json
{
                    "type": "illegal_argument_exception",
                    "reason": "failed to execute script",
                    "caused_by": {
                        "type": "script_exception",
                        "reason": "runtime error",
                        "script_stack": [
                            "f -> f.replace(params.oldFolderPath, params.newFolderPath)); ",
                            "^---- HERE"
                        ],
                        "script": "update-folder-path-in-resources",
                        "lang": "painless",
                        "position": {
                            "offset": 32,
                            "start": 32,
                            "end": 93
                        },
                        "caused_by": {
                            "type": "illegal_argument_exception",
                            "reason": "cannot convert function reference [this::lambda$synthetic$0] to a non-functional interface [java.util.function.UnaryOperator]"
                        }
                    }
```
I'm guessing this is the same bug but not really sure.

**Comment by stu-elastic** [comment]
@TheFireCookie It's the same bug.  It's nice to have a repro for `UnaryOperator`.

In [`LambdaTests.java`](https://github.com/elastic/elasticsearch/blob/master/modules/lang-painless/src/test/java/org/elasticsearch/painless/LambdaTests.java)
```java
    public void testUnaryOperator() {
        assertEquals("doremi", exec("List lst = ['abc', '123']; lst.replaceAll(f -> f.replace('abc', 'doremi')); lst.get(0);"));
        assertEquals("doremi", exec("def lst = ['abc', '123']; lst.replaceAll(f -> f.replace('abc', 'doremi')); lst.get(0);"));
    }
```

**Comment by jdconrad** [comment]
Fixed by https://github.com/elastic/elasticsearch/pull/81698

**Comment by stu-elastic** [comment]
This bug has been fixed in [v7.16.2](https://www.elastic.co/guide/en/elasticsearch/reference/current/release-notes-7.16.2.html).  Thanks for the report.

**Comment by megglos** [comment]
@stu-elastic thanks for the heads-up! You have my thanks for resolving this blocker for us so quickly!
